### PR TITLE
Added ClientSideOperationTimeout class

### PR DESCRIPTION
**Pull request by rozza**
Initial client side operation timeout (CSOT) work. The CSOT class is passed to all operations and currently encapsulate the following timeouts:

  - `timeoutMS` the new optional client side operation timeout
  - `maxTimeMS` the legacy maxTimeMS operation value. Ignored if CSOT is set.
  - `maxCommitTimeMS` the legacy commit timeout. Ignored if CSOT is set.
  - `maxAwaitTimeMS` the getMore await timeout.

This initial work allows for the CSOT to be available to all operations and later work will pass / apply this timeout where required by the Spec.

JAVA-4086

### CODE REVIEW

File: driver-legacy/src/main/com/mongodb/DBCollection.java [hunk 1]
```diff
@@ -2195,6 +2201,19 @@ BsonDocument wrap(final DBObject document, @Nullable final Encoder<DBObject> enc
         }
     }
 
+    private ClientSideOperationTimeout getClientSideOperationTimeout(){
+       return ClientSideOperationTimeouts.create(database.getTimeoutMS());
```
Comment (katcharov) [inline driver-legacy/src/main/com/mongodb/DBCollection.java:2205]: In the spec, there are overrides (Client, DB, Collection, and then (operation or ClientSession)). This seems to be handled by passing the database setting to the collection, allowing for an override via withTimeout. I think JAVA-4064 covers adding it to the client. Is there a test for what is currently implemented, or a Jira that covers whatever is left to finish and test?

**Review by katcharov — CHANGES_REQUESTED** [review]
It's looking good

File: driver-core/src/main/com/mongodb/internal/ClientSideOperationTimeout.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+ * <p>Includes support for the deprecated {@code maxTimeMS} and {@code maxCommitTimeMS} operation configurations</p>
+ */
+public class ClientSideOperationTimeout {
+
+    private final Long timeoutMS;
+    private final long maxAwaitTimeMS;
+
+    // Deprecated operation based operation timeouts
+    private final long maxTimeMS;
+    private final long maxCommitTimeMS;
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/ClientSideOperationTimeout.java:37]: Correct these are in the spec but not handled by this PR. 

This PR does the minimum required to get the Operation based timeouts together.  

When the connection pool CSOT work is done then wait queue timeout can be looked at - its just not generally available at this time. So in some cases the `min` method might be used or other timeouts added to this class.

When commands are issued with write concerns then wTimeoutMS may be set.

File: driver-core/src/test/functional/com/mongodb/internal/operation/ListCollectionsOperationSpecification.groovy [hunk 3]
```diff
@@ -384,40 +386,24 @@ class ListCollectionsOperationSpecification extends OperationFunctionalSpecifica
...
-        enableMaxTimeFailPoint()
-
-        when:
-        executeAsync(operation)
 
-        then:
-        thrown(MongoExecutionTimeoutException)
-
-        cleanup:
-        disableMaxTimeFailPoint()
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-core/src/test/functional/com/mongodb/internal/operation/ListCollectionsOperationSpecification.groovy:420]: Ah good spot - I should have explained. Many many moons ago there were duplicate tests for sync and async like this one and overtime they subtly diverged and hid errors. So using a async flag was adopted, but evidently not all code was updated.

TLDR; I merged the async and sync test together using the:

```
where:
async << [true, false]
```

File: driver-scala/src/main/scala/org/mongodb/scala/MongoDatabase.scala [hunk 4]
```diff
@@ -69,6 +70,28 @@ case class MongoDatabase(private[scala] val wrapped: JMongoDatabase) {
...
+   *
+   * -`null` means that the timeout mechanism for operations will defer to using:
+   * -`waitQueueTimeoutMS`: The maximum wait time in milliseconds that a thread may wait for a connection to become available
+   * -`socketTimeoutMS`: How long a send or receive on a socket can take before timing out.
+   * -`wTimeoutMS`: How long the server will wait for the write concern to be fulfilled before timing out.
+   * -`maxTimeMS`: The time limit for processing operations on a cursor.
+   * See: <a href="https://docs.mongodb.com/manual/reference/method/cursor.maxTimeMS">cursor.maxTimeMS</a>.
+   * -`maxCommitTimeMS`: The maximum amount of time to allow a single `commitTransaction` command to execute.
+   * -`0` means infinite timeout.
+   * -`> 0` The time limit to use for the full execution of an operation.
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-scala/src/main/scala/org/mongodb/scala/MongoDatabase.scala:87]: Scaladoc and Kdoc (Kotlin) have (nicer?) markdown syntax. So have adopted that for those languages.

File: driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/MongoOperationPublisher.java [hunk 5]
```diff
@@ -175,44 +182,54 @@ <D> MongoOperationPublisher<D> withNamespaceAndDocumentClass(final MongoNamespac
...
+
+    MongoOperationPublisher<T> withTimeout(final long timeout, final TimeUnit timeUnit) {
+        long timeoutMS = notNull("timeUnit", timeUnit).toMillis(timeout);
+        if (Objects.equals(getTimeoutMS(), timeoutMS)) {
+            return this;
+        }
+        return new MongoOperationPublisher<>(getNamespace(), getDocumentClass(),
+                getCodecRegistry(), getReadPreference(), getReadConcern(),
+                getWriteConcern(), getRetryWrites(), getRetryReads(), uuidRepresentation,
+                autoEncryptionSettings, getTimeoutMS(), executor);
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/MongoOperationPublisher.java:232]: Good catch!

File: driver-core/src/main/com/mongodb/internal/ClientSideOperationTimeout.java [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+    @Nullable
+    private Timeout timeout;
+
+    public ClientSideOperationTimeout(@Nullable final Long timeoutMS,
+                                      final long maxAwaitTimeMS,
+                                      final long maxTimeMS,
+                                      final long maxCommitTimeMS) {
+        this.timeoutMS = timeoutMS;
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/ClientSideOperationTimeout.java:46]: Yes makes sense. Will do.

File: driver-sync/src/test/unit/com/mongodb/client/internal/MongoCollectionSpecification.groovy [hunk 7]
```diff
@@ -1275,16 +1296,16 @@ class MongoCollectionSpecification extends Specification {
         indexes == []
 
         then:
-        expect operation, isTheSameAs(new ListIndexesOperation(namespace, new BsonDocumentCodec()).retryReads(true))
+        expect operation, isTheSameAs(new ListIndexesOperation(CSOT_NO_TIMEOUT.get(), namespace, new BsonDocumentCodec()).retryReads(true))
         executor.getClientSession() == session
 
         when:
-        execute(listIndexesMethod, session).batchSize(10).maxTime(10, MILLISECONDS).iterator()
+        execute(listIndexesMethod, session).batchSize(10).maxTime(100, MILLISECONDS).iterator()
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-sync/src/test/unit/com/mongodb/client/internal/MongoCollectionSpecification.groovy:1303]: Good Question - its just to reuse the static CSOT suppliers.

File: driver-core/src/test/functional/com/mongodb/internal/operation/DropDatabaseOperationSpecification.groovy [hunk 8]
```diff
@@ -42,47 +44,28 @@ class DropDatabaseOperationSpecification extends OperationFunctionalSpecificatio
...
         when:
-        new DropDatabaseOperation(dbName).execute(getBinding())
+        execute(new DropDatabaseOperation(CSOT_TIMEOUT.get(), dbName, WriteConcern.ACKNOWLEDGED), async)
 
         then:
         !databaseNameExists(dbName)
-    }
-
-
-    def 'should not error when dropping a collection that does not exist asynchronously'() {
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-core/src/test/functional/com/mongodb/internal/operation/DropDatabaseOperationSpecification.groovy:77]: Async and sync tests were merged. (See other comment)

File: driver-core/src/test/functional/com/mongodb/internal/operation/ListIndexesOperationSpecification.groovy [hunk 9]
```diff
@@ -214,43 +214,26 @@ class ListIndexesOperationSpecification extends OperationFunctionalSpecification
...
         then:
         thrown(MongoExecutionTimeoutException)
 
         cleanup:
         disableMaxTimeFailPoint()
-    }
-
-
-    @IgnoreIf({ isSharded() })
-    def 'should throw execution timeout exception from executeAsync'() {
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-core/src/test/functional/com/mongodb/internal/operation/ListIndexesOperationSpecification.groovy:236]: Async and sync tests were merged. (See other comment)

File: driver-sync/src/main/com/mongodb/client/internal/ClientSessionImpl.java [hunk 10]
```diff
@@ -177,9 +181,11 @@ public void abortTransaction() {
                 if (readConcern == null) {
                     throw new MongoInternalException("Invariant violated.  Transaction options read concern can not be null");
                 }
-                delegate.getOperationExecutor().execute(new AbortTransactionOperation(assertNotNull(transactionOptions.getWriteConcern()))
-                                .recoveryToken(getRecoveryToken()),
-                        readConcern, this);
+                delegate.getOperationExecutor().execute(new AbortTransactionOperation(
+                        // TODO - JAVA-4067
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-sync/src/main/com/mongodb/client/internal/ClientSessionImpl.java:185]: Like that - will do

File: driver-legacy/src/main/com/mongodb/DBCollection.java (hunk 1 above)
Comment (rozza) [reply driver-legacy/src/main/com/mongodb/DBCollection.java:2205]: I'm expecting there to be Unified tests for timeouts. 
I haven't added tests as this PR is really to encapsulate what we currently have. But as you spotted in the MongoOperationPublisher I already hid a bug. Will review / ticket.

File: driver-core/src/test/functional/com/mongodb/internal/operation/RenameCollectionOperationSpecification.groovy [hunk 11]
```diff
@@ -35,70 +37,52 @@ import static com.mongodb.ClusterFixture.serverVersionLessThan
...
-        given:
-        getCollectionHelper().insertDocuments(new DocumentCodec(), new Document('documentThat', 'forces creation of the Collection'))
-        assert collectionNameExists(getCollectionName())
-
-        when:
-        executeAsync(new RenameCollectionOperation(getNamespace(), getNamespace()))
-
-        then:
-        thrown(MongoServerException)
-        collectionNameExists(getCollectionName())
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-core/src/test/functional/com/mongodb/internal/operation/RenameCollectionOperationSpecification.groovy:92]: See previous comment

File: driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/MongoOperationPublisher.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (rozza) [reply driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/MongoOperationPublisher.java]: Added `MongoOperationPublisherTest`

File: driver-core/src/main/com/mongodb/internal/ClientSideOperationTimeout.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/ClientSideOperationTimeout.java:48]: Done

File: driver-kotlin-coroutine/src/main/kotlin/com/mongodb/kotlin/client/coroutine/MongoCollection.kt [hunk 12]
```diff
@@ -87,6 +87,26 @@ public class MongoCollection<T : Any>(private val wrapped: JMongoCollection<T>)
     public val writeConcern: WriteConcern
         get() = wrapped.writeConcern
 
+    /** The time limit for the full execution of an operation. */
+    /**
+     * The time limit for the full execution of an operation.
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-kotlin-coroutine/src/main/kotlin/com/mongodb/kotlin/client/coroutine/MongoCollection.kt:91]: Fixed copy pasta.

File: driver-sync/src/main/com/mongodb/client/internal/ClientSessionImpl.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (rozza) [reply driver-sync/src/main/com/mongodb/client/internal/ClientSessionImpl.java:185]: Done

File: driver-kotlin-sync/src/main/kotlin/com/mongodb/kotlin/client/MongoDatabase.kt [hunk 13]
```diff
@@ -53,6 +53,25 @@ public class MongoDatabase(private val wrapped: JMongoDatabase) {
...
 
+    /**
+     * The time limit for the full execution of an operation.
+     *
+     * If not null the following deprecated options will be ignored: `waitQueueTimeoutMS`, `socketTimeoutMS`,
+     * `wTimeoutMS`, `maxTimeMS` and `maxCommitTimeMS`
+     *
+     * -`null` means that the timeout mechanism for operations will defer to using: -`waitQueueTimeoutMS`: The maximum
+     * wait time in milliseconds that a thread may wait for a connection to become available -`socketTimeoutMS`: How
+     * long a send or receive on a socket can take before timing out. -`wTimeoutMS`: How long the server will wait for
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-kotlin-sync/src/main/kotlin/com/mongodb/kotlin/client/MongoDatabase.kt]: Done

File: driver-scala/src/main/scala/org/mongodb/scala/MongoDatabase.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (rozza) [reply driver-scala/src/main/scala/org/mongodb/scala/MongoDatabase.scala]: Done

File: config/checkstyle/checkstyle.xml [hunk 14]
```diff
@@ -122,7 +122,7 @@
 
         <module name="MethodLength"/>
         <module name="ParameterNumber">
-            <property name="max" value="12"/>
+            <property name="max" value="13"/>
```
[the remark this answers is not in the document]
Comment (rozza) [reply config/checkstyle/checkstyle.xml]: Done - it was AggregateIterableImpl

**Review by katcharov — APPROVED** [review]
LGTM!
